File: server/models/Memory.js
const mongoose = require('mongoose');

const MemorySchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  type: {
    type: String,
    enum: ['fact', 'preference', 'event', 'insight', 'profile', 'summary'],
    required: true,
  },
  content: {
    type: String,
    required: true,
  },
  key: {
    type: String,
  },
  value: {
    type: String,
  },
  category: {
    type: String,
    enum: ['personal', 'education', 'career', 'health', 'tech', 'lifestyle', 'social', 'general'],
    default: 'general',
  },
  chatId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Chat',
  },
  source: {
    type: String,
    enum: ['chat', 'goal', 'habit', 'manual'],
    required: true,
  },
  sourceId: {
    type: mongoose.Schema.Types.ObjectId,
  },
  importance: {
    type: Number,
    min: 1,
    max: 10,
    required: true,
    default: 5,
  },
  tags: [{
    type: String,
  }],
  expiresAt: {
    type: Date,
  }
}, {
  timestamps: true
});

// Indexes
MemorySchema.index({ userId: 1 });
MemorySchema.index({ importance: -1 });
MemorySchema.index({ createdAt: -1 });
MemorySchema.index({ userId: 1, key: 1 });
MemorySchema.index({ userId: 1, category: 1 });
MemorySchema.index({ userId: 1, type: 1 });

module.exports = mongoose.model('Memory', MemorySchema);

File: flutter_client/lib/features/memory/models/memory.dart
class Memory {
  final String id;
  final String type;
  final String content;
  final String? key;
  final String? value;
  final String category;
  final String? chatId;
  final String source;
  final String? sourceId;
  final int importance;
  final List<String> tags;
  final DateTime? expiresAt;
  final DateTime createdAt;

  Memory({
    required this.id,
    required this.type,
    required this.content,
    this.key,
    this.value,
    required this.category,
    this.chatId,
    required this.source,
    this.sourceId,
    required this.importance,
    this.tags = const [],
    this.expiresAt,
    required this.createdAt,
  });

  factory Memory.fromJson(Map<String, dynamic> json) {
    return Memory(
      id: json['_id'] as String,
      type: json['type'] as String? ?? 'general',
      content: json['content'] as String,
      key: json['key'] as String?,
      value: json['value'] as String?,
      category: json['category'] as String? ?? 'general',
      chatId: json['chatId'] as String?,
      source: json['source'] as String? ?? 'manual',
      sourceId: json['sourceId'] as String?,
      importance: json['importance'] as int? ?? 5,
      tags: (json['tags'] as List<dynamic>?)?.map((e) => e.toString()).toList() ?? [],
      expiresAt: json['expiresAt'] != null ? DateTime.parse(json['expiresAt'] as String) : null,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt'] as String) : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'content': content,
      if (key != null) 'key': key,
      if (value != null) 'value': value,
      'category': category,
      if (chatId != null) 'chatId': chatId,
      'source': source,
      if (sourceId != null) 'sourceId': sourceId,
      'importance': importance,
      'tags': tags,
      if (expiresAt != null) 'expiresAt': expiresAt!.toIso8601String(),
    };
  }
}
